nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -609,6 +620,74 @@ in
       systemd.targets.zfs.wantedBy = [ "multi-user.target" ];
     })
 
+    (mkIf (cfgZfs.enabled && cfgExpandOnBoot != false) {
+      systemd.services."zpool-expand@" = {
+        description = "Expand ZFS pools";
+        after = [ "zfs.target" ];
+
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+        };
+
+        scriptArgs = "%i";
+        path = [ pkgs.gawk cfgZfs.package ];
+
+        # Note this script attempts to resize more than just devices.
+        #
+        # It will also try to `online -e` the pool itself and every
+        # vdev. This will print otherwise harmless errors.
+        #
+        # ZFS has no way of enumerating _just_ devices in a pool in a way
+        # that `zpool online -e` supports.
+        #
+        # See: https://github.com/openzfs/zfs/issues/12505
+        script = ''
+          pool=$1
+
+          echo "Expanding all devices for $pool"
+
+          for device in $(zpool list -vH "$pool" | awk '($10 = "ONLINE") {print $1;}'); do
+            echo "Attempting to expand $device of $pool"
+            zpool online -e "$pool" "$device" || true
```

Comment:
Is there a way to filter the devices that are collected? Maybe remove anything that matches the name of a pool, as well as the `raidz[1-9]` format? Or, remove anything that doesn't match `(ata|nvme|usb|.....)-.*`?

Alternatively (instead of `|| true`), print out a message assuaging any concerns that this is a fatal error or will otherwise affect the system's stability?

Maybe even an additional option specifying the devices to expand instead of manually detecting them...?

Reply:
> Is there a way to filter the devices that are collected? Maybe remove anything that matches the name of a pool, as well as the `raidz[1-9]` format? Or, remove anything that doesn't match `(ata|nvme|usb|.....)-.*`?

Unfortunately I don't think there is a correct way to filter out non-devices. There may be ways with `zdb`, but zdb's interface is unstable and not to be scripted against. See the ticket for various approaches I took and failed at.

> Alternatively (instead of `|| true`), print out a message assuaging any concerns that this is a fatal error or will otherwise affect the system's stability?

Yeah, I think that is a good idea.
 
> Maybe even an additional option specifying the devices to expand instead of manually detecting them...?

I am not keen on this. Devices can change names between boots, and many of their identifiers aren't incredibly stable. Some of the paths like `by-id` fix this, but I don't really want to walk down this road.

Reply:
I have found a way to make it only attempt to expand actual devices, in the new commit. Take another look?

Reply:
A much better solution (if a bit... gross).